hal/armv7m/mpu.c

```diff
@@ -0,0 +1,322 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Phoenix-RTOS loader
+ *
+ * MPU routines
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Gerard Swiderski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <lib/lib.h>
+#include <lib/log.h>
+#include <lib/errno.h>
+#include <hal/hal.h>
+#include <syspage.h>
+#include "mpu.h"
+
+
+enum { mpu_fError = 0x01, mpu_fConfigured = 0x04, mpu_fValid = 0x08 };
+
+
+struct {
+	syspage_hal_t hal;
+	u8 mapsCnt;
+	u8 regCnt;
+	u8 regMax;
+	u8 flags;
+} mpu_common;
+
+
+/* Binary count trailing zeros */
+__attribute__((always_inline)) static inline u8 _ctz(u32 val)
+{
+	asm volatile(" \
+		rbit %0, %0; \
+		clz %0, %0"
+		: "+l"(val));
+	return val;
+}
+
+
+static void mpu_attrPrint(u32 attr, int enabled)
+{
+	static const char *strAP[8] = { "na/na", "rw/na", "rw/ro", "rw/rw", "reserved", "ro/ro", "ro/ro" };
+	static const char strYesNo[2] = { 'n', 'y' };
+
+	/* Enabled */
+	lib_printf("%c   ", strYesNo[enabled != 0]);
+
+	/* eXec Never */
+	lib_printf("%c   ", strYesNo[(attr >> 12) & 1ul]);
+
+	/* Access Permission */
+	lib_printf("%-12s", strAP[(attr >> 8) & 0x7ul]);
+
+	/* TEX */
+	lib_printf("%c", '0' + ((attr >> 5) & 1ul));
+	lib_printf("%c", '0' + ((attr >> 4) & 1ul));
+	lib_printf("%c  ", '0' + ((attr >> 3) & 1ul));
+
+	/* Shareable */
+	lib_printf("%c  ", strYesNo[(attr >> 2) & 1ul]);
+
+	/* Cacheeable */
+	lib_printf("%c  ", strYesNo[(attr >> 1) & 1ul]);
+
+	/* Bufferable */
+	lib_printf("%c\n", strYesNo[attr & 1ul]);
+}
+
+
+/* Describe content of rbar & rasr registers */
+static void mpu_regionPrint(const char *name, u32 rbar, u32 rasr)
+{
+	unsigned int bit;
+	u32 srSize, srBase = rbar & (0x7fffffful << 5);
+	u32 attr = (rasr >> 16) & 0x173f;
+	u8 sizeBit = ((rasr >> 1) & 0x1fu) + 1;
+	u8 srdMask = rasr >> 8;
+	u8 region = rbar & 0xfu;
+
+	if (srdMask != 0) {
+		srSize = 1u << (sizeBit - 3);
+
+		for (bit = 0; bit < 8; bit++) {
+			if ((srdMask & (1u << bit)) == 0) {
+				lib_printf("%-10s %d%-5s %d%-8s 0x%08x%4s 0x%08x%4s",
+					name, region, "", bit, "", srBase, "", srBase + srSize - 1, "");
+				mpu_attrPrint(attr, (rasr & 1));
+			}
+
+			srBase += srSize;
+		}
+	}
+	else {
+		lib_printf("%-10s %d%-5s %-9s 0x%08x%4s 0x%08x%4s",
+			name, region, "", "all subregions", srBase, "", srBase + (1ul << sizeBit) - 1, "");
+		mpu_attrPrint(attr, (rasr & 1));
+	}
+}
+
```

Comment:
This function is redundant. All information about maps should be displayed by map command belonging to hardware independent part.

Reply:
Which command should display hardware dependent setup (e.g. region setup?) like on the screenshot attached in PR introduction above ?

Reply:
Removed, no more printing in hal.